sdk/azcore/runtime/policy_bearer_token.go

```diff
@@ -103,9 +112,37 @@ func (b *BearerTokenPolicy) Do(req *policy.Request) (*http.Response, error) {
 
 	if res.StatusCode == http.StatusUnauthorized {
 		b.mainResource.Expire()
-		if res.Header.Get("WWW-Authenticate") != "" && b.authzHandler.OnChallenge != nil {
-			if err = b.authzHandler.OnChallenge(req, res, b.authenticateAndAuthorize(req)); err == nil {
-				res, err = req.Next()
+		if res.Header.Get(shared.HeaderWWWAuthenticate) != "" {
+			caeChallenge, parseErr := parseCAEChallenge(res)
+			if parseErr != nil {
+				return res, parseErr
+			}
+			switch {
+			case caeChallenge != nil:
+				tro := policy.TokenRequestOptions{
+					Claims: caeChallenge.params["claims"],
+					Scopes: b.scopes,
+				}
+				if err = b.authenticateAndAuthorize(req)(tro); err == nil {
+					res, err = req.Next()
+				}
+			case b.authzHandler.OnChallenge != nil:
+				if err = b.authzHandler.OnChallenge(req, res, b.authenticateAndAuthorize(req)); err == nil {
+					if res, err = req.Next(); err == nil && res.StatusCode == http.StatusUnauthorized {
```

Comment:
This looks to be duplicating earlier code. Can it be simplified to remove the duplication?

Reply:
I can reduce duplication but I'm uncertain the result will be simpler. At first glance I think it needs recursion. I'll give it a shot

Reply:
Well, it turned out recursive but not terrible. What do you think?

Reply:
Looks good!